src/rust/ide/ast/impl/src/crumbs.rs

```diff
@@ -511,15 +988,292 @@ mod tests {
+
+
+    // === TextLineFmt ===
+
     #[test]
-    fn iterate_infix() {
-        let sum = crate::Infix::from_vars("foo", "+", "bar");
-        let (larg,opr,rarg) = sum.iter_subcrumbs().expect_tuple();
-        assert_eq!(larg, InfixCrumb::LeftOperand);
-        assert_eq!(opr,  InfixCrumb::Operator);
-        assert_eq!(rarg, InfixCrumb::RightOperand);
+    fn text_line_fmt_crumb() {
+        let expr   = SegmentExpr { value : Some(Ast::var("foo")) };
+        let text   = vec![SegmentFmt::SegmentExpr(expr)];
+        let ast    = Ast::text_line_fmt(text);
+        let crumbf = |crumb| Crumb::TextLineFmt(crumb);
+        let bar    = Ast::var("bar");
+        let crumb  = TextLineFmtCrumb{segment_index:0};
+
+        assert_eq!(ast.repr(), "'`foo`'");
+        assert_eq!(get(crumbf,&ast,crumb).unwrap().repr(),  "foo");
+        assert_eq!(set(crumbf,&ast,crumb,bar).unwrap().repr(), "'`bar`'");
+    }
+
+    #[test]
+    fn iterate_text_line_fmt() {
+        let expr1  = SegmentExpr { value : Some(Ast::var("foo")) };
+        let expr2  = SegmentExpr { value : Some(Ast::var("bar")) };
+        let text   = vec![SegmentFmt::SegmentExpr(expr1),SegmentFmt::SegmentExpr(expr2)];
+        let ast    = Ast::text_line_fmt(text);
+
+        let (segment1,segment2) = ast.iter_subcrumbs().expect_tuple();
+
+        assert_eq!(segment1, Crumb::TextLineFmt(TextLineFmtCrumb{segment_index:0}));
+        assert_eq!(segment2, Crumb::TextLineFmt(TextLineFmtCrumb{segment_index:1}));
     }
 
+    // === TextBlockFmt ===
+
+    #[test]
+    fn text_block_fmt_crumb() {
+        let empty_lines = default();
+        let expr        = SegmentExpr { value : Some(Ast::var("foo")) };
+        let text        = vec![SegmentFmt::SegmentExpr(expr)];
+        let line1       = TextBlockLine{empty_lines,text};
+
+        let empty_lines = default();
+        let expr        = SegmentExpr { value : Some(Ast::var("bar")) };
+        let text        = vec![SegmentFmt::SegmentExpr(expr)];
+        let line2       = TextBlockLine{empty_lines,text};
+
+        let lines       = vec![line1,line2];
+        let ast         = Ast::text_block_fmt(lines);
+        let qux         = Ast::var("qux");
+        let baz         = Ast::var("baz");
+
+        let crumbf = |crumb| Crumb::TextBlockFmt(crumb);
+        assert_eq!(ast.repr(), "'''\n`foo`\n`bar`");
+
+        let crumb1 = TextBlockFmtCrumb {text_line_index:0, segment_index:0};
+        let crumb2 = TextBlockFmtCrumb {text_line_index:1, segment_index:0};
+
+        assert_eq!(get(crumbf,&ast,crumb1).unwrap().repr(), "foo");
+        assert_eq!(get(crumbf,&ast,crumb2).unwrap().repr(), "bar");
+
+        assert_eq!(set(crumbf,&ast,crumb1,qux).unwrap().repr(),"'''\n`qux`\n`bar`");
+        assert_eq!(set(crumbf,&ast,crumb2,baz).unwrap().repr(),"'''\n`foo`\n`baz`");
+    }
+
+    #[test]
+    fn iterate_text_block_fmt() {
+        let empty_lines = default();
+        let expr        = SegmentExpr { value : Some(Ast::var("foo")) };
```

Comment:
As before — should also check the case with presence of other segment types. 
The logic of iterating is here largely the logic of filtering.